src/devices/bus/a2bus/a2frob.cpp

```diff
@@ -0,0 +1,430 @@
+
+	virtual void write_cnxx(u8 offset, u8 data) override
+	{ if (!m_frob_apple_control) m_frob_mem[(offset & 0xff)|(m_frob_page << 8)] = data; }
+
+	virtual bool take_c800() override { return false; }
+
+	u8 m_frob_mem[0x1000] = { 0 };  // 4k frob memory
+
+	u8 read_frob_mem(offs_t offset) { return !m_frob_apple_control ? 0 : m_frob_mem[offset & 0xfff]; }
+
+	void write_frob_mem(offs_t offset, u8 data) { m_frob_mem[offset & 0xfff] = data; }
+
+public:
+	void a2600_write_fff0(offs_t offset, u8 data)
+	{
+		m_byte_for_apple = data;
+		m_byte_waiting_flag_apple = 1;
+	}
+
+	u8 a2600_read_fff2(offs_t offset)
+	{
+		if (!machine().side_effects_disabled()) m_byte_waiting_flag_vcs = 0;
+		return m_byte_for_vcs;
+	}
+
+	u8 a2600_read_fff1(offs_t offset)
+	{
+		return (m_byte_waiting_flag_apple ? 0 : 1) << 7 | (m_byte_waiting_flag_vcs << 6);
+	}
+
+protected:
+
+	void switch_A_w(uint8_t data);
+	uint8_t switch_A_r();
+	void switch_B_w(uint8_t data);
+	void irq_callback(int state);
+	uint16_t a2600_read_input_port(offs_t offset);
+	uint8_t a2600_get_databus_contents(offs_t offset);
+	void a2600_tia_vsync_callback(uint16_t data);
+
+	required_shared_ptr<uint8_t> m_riot_ram;
+	required_device<tia_video_device> m_tia;
+	required_device<m6507_device> m_maincpu;
+	required_device<mos6532_device> m_riot;
+	required_device<vcs_control_port_device> m_joy1;
+	required_device<vcs_control_port_device> m_joy2;
+	required_device<screen_device> m_screen;
+
+	u8 m_frob_apple_control = 0;  // 0 means that it's under apple control
+	u8 m_byte_waiting_flag_apple = 0;
+	u8 m_byte_waiting_flag_vcs = 0;
+	u8 m_byte_for_vcs = 0;
+	u8 m_byte_for_apple = 0;
+	u8 m_bidirectional_active = 0;
+	u8 m_frob_page = 0;
+	u8 m_frob_control_reg = 0;
+
+private:
+	uint16_t m_current_screen_height = 0U;
+	XTAL m_xtal;
+
+};
+
+a2600_frob_base_device::a2600_frob_base_device(machine_config const &mconfig, char const *tag, device_t *owner, u32 clock) :
+	a2600_frob_base_device(mconfig, A2BUS_FROB, tag, owner, clock)
+{
+}
+
+
+u8 a2600_frob_base_device::read_c0nx(u8 offset)
+{
+	// offset 0 is read from frob status  (bits 7 = write ok, 6 = read ok) (other bits will float)
+	switch (offset & 0x1)
+	{
+		case 0:
+			return ((m_byte_waiting_flag_vcs ? 0 : 1) << 7) |
+					(m_byte_waiting_flag_apple << 6);
+			break;
+		case 1:
+			if (!machine().side_effects_disabled())
+			{
+				m_byte_waiting_flag_apple = 0; // clear byte waiting for apple
+			}
+			return m_byte_for_apple;
+		default:
+			return 0;
+	}
+}
+
+void a2600_frob_base_device::write_c0nx(u8 offset, u8 data)
+{
+	// offset 0 is write to frob control register
+	// offset 1 is write to bidirectional register
+	switch (offset & 0x1)
+	{
+		case 0:
+			m_frob_control_reg = data;
+			m_frob_apple_control = BIT(data, 4);
+			m_bidirectional_active = BIT(data, 5);
+			m_frob_page = BIT(data, 0, 4);
+//          printf("m_frob_apple_control = %x  m_frob_page = %x\n",m_frob_apple_control, m_frob_page);
+			break;
+		case 1:
```

Comment:
Use logmacro.h helpers for configurable logging, not commented prints.  Commented code always rots.